SSA/Projects/FullyHomomorphicEncryption/Basic.lean

```diff
@@ -81,72 +142,240 @@ abbrev R := (ZMod q)[X] ⧸ (Ideal.span {f q n})
 -- Coefficients of `a : R' q n` are `a\_i : Zmod q`.
 -- TODO: get this from mathlib
 
-/-- Canonical epimorphism `ZMod q[X] ->*+ R q n` -/
+/-- Canonical epimorphism / quotient map: `ZMod q[X] ->*+ R q n` -/
 abbrev R.fromPoly {q n : Nat} : (ZMod q)[X] →+* R q n := Ideal.Quotient.mk (Ideal.span {f q n})
 
-noncomputable abbrev R.zero {q n : Nat} : R q n := R.fromPoly 0
-noncomputable abbrev R.one {q n : Nat} : R q n := R.fromPoly 1
+/-- fromPoly, the canonical epi from `ZMod q[X] →*+ R q n` is surjective -/
+theorem R.surjective_fromPoly (q n : ℕ) : Function.Surjective (R.fromPoly (q := q) (n := n)) := by
+  simp[R.fromPoly]
+  apply Ideal.Quotient.mk_surjective
+
+private noncomputable def R.representative' : R q n → (ZMod q)[X] := Function.surjInv (R.surjective_fromPoly q n)
```

Comment:
What's the point in this when you have `representative`. There's never a reason to use this one, you just know less things about it and it could even be equal to `representative`

Reply:
The point of this is to use it in order to define `representative`, because we want to the one that has degree < deg f. We could just inline the definition, I personally find it more readable this way, but I can easily be convinced otherwise.